version: "3.7"

services:
  nginx:
    container_name: pypi_nginx
    hostname: nginx
    image: nginx:1.18.0-alpine
    privileged: true
    user: ${CURRENT_UID}
    restart: unless-stopped
    networks:
      pypimanager_net:
        ipv4_address: 192.199.41.2
    ports:
      # 8898为pypiserver的端口
      - 8898:8898
      # 8897为pypimanager的WEB端口
      - 8897:8897
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - $PWD/pypimanager-nginx/conf:/etc/nginx
      - $PWD/pypimanager-nginx/cache:/data/nginx/cache
      - $PWD/pypimanager-nginx/logs:/var/log/nginx
      - $PWD/pypiserver/packages:/data/packages
      - $PWD/pypimanager-frontend/dist:/pypimanager-frontend
    depends_on:
      - server
      - manager
  server:
    container_name: pypi_server
    image: pypiserver/pypiserver:v1.4.2
    restart: unless-stopped
    privileged: true
    user: ${CURRENT_UID}
    networks:
      pypimanager_net:
        ipv4_address: 192.199.41.3
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - $PWD/pypiserver/packages:/data/packages
      - $PWD/pypiserver/logs:/logs
      - $PWD/pypiserver/pypipasswd:/pypipasswd
    command: -v --log-file=/logs/pypiserver.log -P /pypipasswd -a update -r /data/packages -o
  db:
    image: mysql:5.7
    container_name: pypi_mysql
    restart: unless-stopped
    user: ${CURRENT_UID}
    networks:
      pypimanager_net:
        ipv4_address: 192.199.41.4
    ports:
      - 3306:3306
    volumes:
      - /etc/localhost:/etc/localhost:ro
      - $PWD/pypimanager-mysql/data:/var/lib/mysql
      - $PWD/pypimanager-mysql/conf/my.cnf:/etc/mysql/my.cnf
      - $PWD/pypimanager-mysql/logs:/var/log/mysql
    environment:
      # 请务必修改密码
      MYSQL_ROOT_PASSWORD: "123456"
    command: --default-authentication-plugin=mysql_native_password
    entrypoint: bash -c "chown -R mysql:mysql /var/log/mysql && chmod 1777 /var/log/mysql && exec /entrypoint.sh mysqld"
  manager:
    container_name: pypi_manager
    image: toddlerya/py_env:3.7.9-buster
    restart: unless-stopped
    privileged: true
    user: ${CURRENT_UID}
    command: ["/sbin/init"]
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:5000"]
      interval: 300s
      timeout: 20s
      retries: 3
    networks:
      pypimanager_net:
        ipv4_address: 192.199.41.5
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - $PWD/pypiserver/packages:/pypi_packages
      - $PWD/pypiserver/recycle_bin:/pypi_packages_recycle_bin
      - $PWD/pypimanager-nginx/logs:/pypimanager-nginx/logs
      - $PWD/pypimanager-backend/pypimanager_backend:/home/pypimanager_backend
    depends_on:
      - db
      - server

networks:
  pypimanager_net:
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 192.199.41.0/16
